[Intents] Add Intents framework from Xcode 8 beta2

dalexsoto opened the pull request:
This PR adds two things

A generator fix to trampolines because we were generating invalid code 

- Bad: `del ( Runtime.GetNSObject<Intents.INIntentResolutionResult'1[[Intents.INRestaurantGuest, temp, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null]]> (obj));`
- Good: `del ( Runtime.GetNSObject<global::Intents.INIntentResolutionResult<global::Intents.INRestaurantGuest>> (obj));`

The Intents framework bindings, I just pushed the Beta 2 commit merged (updated) with Beta 1 stuff since I could not land this before beta 2 came out

Introspection tests passes. I will submit the UI part once this PR is accepted since it depends on it

spouliot commented:
Generator changes are correct - and it might be better to commit them separately since it should not be squashed with the bindings (and those are large and will likely need a squash).

spouliot commented on tests/introspection/ApiSelectorTest.cs:
```diff
@@ -165,6 +165,13 @@ protected virtual bool Skip (Type type, string selectorName)
 				}
 				break;
+			case "INPriceRange":
+			switch (selectorName) {
```
indentation ->

xamarin-release-manager commented:
Build success

spouliot commented:
About 50% reviewed

- Many enums are missing their availability attributes. It's kind of implied by the API that consume them but it's nice for documentation (and testing).

- Two letter acronyms must remain in capitals, e.g. `INRadioType` AM, FM and HD (but Dab remains as is)

- Since the framework is not included in tvOS and watchOS you can remove, everywhere both: 
	`[Unavailable (PlatformName.WatchOS)]`
	`[Unavailable (PlatformName.TvOS)]`
	
- It's likely a 64bits only framework on macOS, @chamons can confirm that, and will need to be specified in the availability attributes;

- Whenever possible move availability on type, e.g. `INBookRestaurantReservationIntentHandling` only some members have them - but it could be there only once for all members (everything covered without duplication). Right now it's not clear, for some protocols, if macOS is supported (or not).

- `INIntentIdentifiers` can be bound directly as an enum (with `[Field]`) to avoid duplicating them;

- Note: there are many `identifier`" selectors in different types (all but one are getters). We need to know if all of them maps to this enum and if the setter can be set of other values (i.e. unit tests)

- xtro should help spot more things, like missing or extra :)

spouliot commented:
Introspection failed on the Mac (bots runs 10.11) because of the lack of attributes (mentioned above).

spouliot commented:
not much more :)

- there are some `identifiers` (plural) selectors, that returns an array, to consider too (wrt enum);

- it _might_ be worth to hide things like:

`NSNumber EnableFan { get; }`

which is _likely_ used to provide a `bool?` in objc. Worth a comment in `INSetClimateSettingsInCarIntent` (several) and `INSetDefrosterSettingsInCarIntent`

- `// [Unavailable (PlatformName.iOSAppExtension)]` <- better use `[Advice ("This API is not allowed in extensions")]` right now

chamons commented:
Yep. 64-only. 

/System/Library/Frameworks/Intents.framework/Intents: Mach-O 64-bit dynamically linked shared library x86_64

dalexsoto commented:
Implemented feedback.

I added the attributes to classes and protocols just not to enums yet, I need to setup a VM with 10.12 in order to run introspection tests so I can completely verify that the attributes are correct. Right know they make sense but I could be proven wrong 😄 

> Note: there are many identifier" selectors in different types (all but one are getters). We need to know if all of them maps to this enum and if the setter can be set of other values (i.e. unit tests)
> there are some identifiers (plural) selectors, that returns an array, to consider too (wrt enum);

Just one case matches the identifier enum values, the other ones are either custom identifiers or supplied by the framework according to docs.

> it might be worth to hide things like: `NSNumber EnableFan { get; }`

I completely agree and got a list here https://gist.github.com/dalexsoto/44e363dbeeac96e67849e7c56d9fcca0 added a TODO so I do not forget

xamarin-release-manager commented:
Build failure

xamarin-release-manager commented:
Build failure

spouliot commented:
Interesting, once bindings catch up with Apple we should review all `NS_REFINED_FOR SWIFT` to see if we can apply similar _refinements_.

spouliot commented:
_I need to setup a VM with 10.12_ that can wait a bit, we'll do a pass later specific for macOS. Right now follow the headers and the bot (10.11) results.

spouliot commented:
Failure is unrelated to commit. It's good to merge!
